Game/Scripts/BossShieldAttackScript.cpp

```diff
@@ -0,0 +1,252 @@
+#include "StdAfx.h"
+#include "BossShieldAttackScript.h"
+
+//#include "Application.h"
+//#include "Modules/ModuleScene.h"
+//#include "Scene/Scene.h"
+
+#include "Components/ComponentScript.h"
+#include "Components/ComponentTransform.h"
+#include "Components/ComponentRigidBody.h"
+
+#include "../Scripts/EnemyClass.h"
+#include "../Scripts/BossShieldScript.h"
+#include "../Scripts/EnemyDroneScript.h"
+#include "../Scripts/EnemyVenomiteScript.h"
+#include "../Scripts/HealthSystem.h"
+
+REGISTERCLASS(BossShieldAttackScript);
+
+BossShieldAttackScript::BossShieldAttackScript() : Script(), bossShieldObject(nullptr), isShielding(false),
+	shieldingTime(0.0f), shieldingMaxTime(20.0f), triggerShieldAttackCooldown(false), shieldAttackCooldown(0.0f),
+	shieldAttackMaxCooldown(50.0f), triggerEnemySpawning(false), enemiesToSpawnParent(nullptr),
+	enemySpawnTime(0.0f), enemyMaxSpawnTime(2.0f), battleArenaAreaSize(nullptr), healthSystemScript(nullptr)
+{
+	REGISTER_FIELD(shieldingMaxTime, float);
+	REGISTER_FIELD(shieldAttackMaxCooldown, float);
+
+	REGISTER_FIELD(bossShieldObject, BossShieldScript*);
+
+	REGISTER_FIELD(enemyMaxSpawnTime, float);
+
+	REGISTER_FIELD(enemiesToSpawnParent, GameObject*);
+
+	REGISTER_FIELD(battleArenaAreaSize, ComponentRigidBody*);
+}
+
+void BossShieldAttackScript::Start()
+{
+	shieldingTime = shieldingMaxTime;
+
+	if (enemiesToSpawnParent == nullptr)
+	{
+		return;
+	}
+
+	enemiesReadyToSpawn.reserve(enemiesToSpawnParent->GetChildren().size());
+	enemiesNotReadyToSpawn.reserve(enemiesToSpawnParent->GetChildren().size());
+
+	for (GameObject* enemyToSpawn : enemiesToSpawnParent->GetChildren())
+	{
+		enemiesReadyToSpawn.push_back(enemyToSpawn);
+	}
+
+	healthSystemScript = owner->GetComponent<HealthSystem>();
+}
+
+void BossShieldAttackScript::Update(float deltaTime)
+{
+	if (bossShieldObject->WasHitBySpecialTarget())
+	{
+		shieldingTime = 0.0f;
+		bossShieldObject->DisableHitBySpecialTarget();
+	}
+
+	ManageShield(deltaTime);
+	ManageEnemiesSpawning(deltaTime);
+
+	ManageRespawnOfEnemies();
+}
+
+void BossShieldAttackScript::TriggerShieldAttack()
+{
+	LOG_INFO("The shield attack was triggered");
+
+	bossShieldObject->ActivateShield();
+	healthSystemScript->SetIsImmortal(true);
+
+	isShielding = true;
+	shieldAttackCooldown = shieldAttackMaxCooldown;
+
+	triggerEnemySpawning = true;
+}
+
+bool BossShieldAttackScript::CanPerformShieldAttack() const
+{
+	return shieldAttackCooldown <= 0.0f && !isShielding;
```

Comment:
```suggestion
	return shieldAttackCooldown <= 0.0f && !IsAttacking();
```
In the unlikely case we add more attacks

Reply:
Unlikely but eh, its a quick change, good catch!